apps/pipelines/nodes/nodes.py

```diff
@@ -51,32 +60,68 @@ def _process(self, state: PipelineState) -> PipelineState:
         output = llm.invoke(state["messages"][-1], config=self._config)
         return output.content
 
-    def get_chat_model(self):
+    def get_llm_service(self):
         from apps.service_providers.models import LlmProvider
 
-        provider = LlmProvider.objects.get(id=self.llm_provider_id)
         try:
-            service = provider.get_llm_service()
-            return service.get_chat_model(self.llm_model, self.llm_temperature)
+            provider = LlmProvider.objects.get(id=self.llm_provider_id)
+            return provider.get_llm_service()
         except LlmProvider.DoesNotExist:
             raise PipelineNodeBuildError(f"LLM provider with id {self.llm_provider_id} does not exist")
         except ServiceProviderConfigError as e:
             raise PipelineNodeBuildError("There was an issue configuring the LLM service provider") from e
 
+    def get_chat_model(self):
+        return self.get_llm_service().get_chat_model(self.llm_model, self.llm_temperature)
+
 
-class CreateReport(LLMResponse):
-    __human_name__ = "Create a report"
+class LLMResponseWithPrompt(LLMResponse):
+    __human_name__ = "LLM response with prompt"
 
+    source_material_id: SourceMaterialId | None = None
     prompt: str = (
         "Make a summary of the following text: {input}. "
         "Output it as JSON with a single key called 'summary' with the summary."
     )
 
     def _process(self, state: PipelineState) -> PipelineState:
-        chain = PromptTemplate.from_template(template=self.prompt) | super().get_chat_model()
-        output = chain.invoke(state["messages"][-1], config=self._config)
+        prompt = PromptTemplate.from_template(template=self.prompt)
+        context = self._get_context(state, prompt)
+        chain = prompt | super().get_chat_model()
+        output = chain.invoke(context, config=self._config)
         return output.content
 
+    def _get_context(self, state: PipelineState, prompt: PromptTemplate):
+        session = state["experiment_session"]
+        context = {"input": state["messages"][-1]}
+
+        if "source_material" in prompt.input_variables and self.source_material_id is None:
+            raise PipelineNodeBuildError("No source material set, but the prompt expects it")
```

Comment:
can we validate this when saving the node as well?

Reply:
We can, but we don't currently have a way of returning validations to the frontend. I can add that as a separate story.

Reply:
https://github.com/dimagi/open-chat-studio/issues/627